### PR TITLE
Return error if env file contains non-ascii or utf8 bytes (for windows)

### LINKED ISSUE #26179: Docker run env-file parameter crashes on windows with non-ascii encoding
Using the `--env-file` parameter on windows causes an error `container XXX encountered an error during CreateProcess failed in Win32: The data is invalid. (0xd)`.

**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      windows/amd64
```

**Output of `docker info`:**

```
Containers: 15
 Running: 0
 Paused: 0
 Stopped: 15
Images: 368
Server Version: 1.12.0
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: nat transparent null overlay
Swarm: inactive
Security Options:
Kernel Version: 10.0 14300 (14300.1045.amd64fre.rs1_release_svc.160705-1059)
Operating System: Windows Server 2016 Datacenter Technical Preview 5
OSType: windows
Architecture: x86_64
CPUs: 1
Total Memory: 3.5 GiB
Name: matt-dckr-2016
ID: CVFQ:6K7E:MZMF:KJUS:QZQ2:YDVU:BRGY:4ZQ3:32MQ:5IXI:C4RY:GV76
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Windows Server 2016 TP 5 VM on Azure, created via the `Deploy to Azure` button on https://msdn.microsoft.com/en-au/virtualization/windowscontainers/quick_start/quick_start_windows_server

**Steps to reproduce the issue:**
1. `echo FOO=BAR > env.list`
2. `docker run --env-file env.list --rm microsoft/windowsservercore cmd /c set`

**Describe the results you received:**
An error message:
```
C:\Program Files\Docker\docker.exe: Error response from daemon: container 5f2e0ccb28f1523aa63f50ad6d6e4b4a59086e1ef23fb43c1d3a
0a880f86b346 encountered an error during CreateProcess failed in Win32: The data is invalid. (0xd) extra info: {"ApplicationNa
me":"","CommandLine":"cmd /c set","WorkingDirectory":"C:\\","Environment":{"\u0000":"","F\u0000O\u0000O\u0000":"\u0000B\u0000A
\u0000R\u0000\r\u0000"},"EmulateConsole":false,"CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":true,"Console
Size":[0,0]}.
```

**Describe the results you expected:**
A list of environment variables that includes the variable `FOO`.

**Additional information you deem important (e.g. issue happens only occasionally):**
This also happens in the latest docker nightly:
```
Client:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7
 Git commit:   ec82819
 Built:        Wed Aug 31 01:48:34 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7
 Git commit:   ec82819
 Built:        Wed Aug 31 01:48:34 2016
 OS/Arch:      windows/amd64
```

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to address the issue raised in #26179 where an env file with non-ascii or utf8 bytes will crash on windows platform.

The issue is two-fold:
- Windows will adds a BOM mark at the begining with Notepad as the editor
- Non-utf8 bytes can not be handled by env file parser.

**\- How I did it**

This fix removes utf8 BOM marker if exists so that utf8 encoded env file could be processed.

This fix also returns an error (instead of a runtime CreateProcess crash) if env file contains non-utf8 bytes, thus giving users better experiences.

**\- How to verify it**

Additional test cases has been added in unit tests.

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #26179.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

**Comment by yongtang** [comment]
cc @jhowardmsft @matt-richardson please take a look.

**Comment by matt-richardson** [comment]
### Unicode (UTF16?) ###
```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding unicode
C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
```
-> Success, i think?

### BigEndianUnicode ###
```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding bigendianunicode
C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
```
-> Success, i think?

### UTF8 ###
```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf8
.\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
...
FOO=BAR
...
```
-> success

### UTF7 ###
```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf7
.\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
...
```
-> failure - no FOO variable

### UTF32 ###
```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf32
C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
```
-> Success, i think?

### ASCII ###
 ```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding ascii
.\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
...
FOO=BAR
...
```
-> success

### OEM ###
 ```
"FOO=BAR" | Out-File -FilePath .\env.list -Encoding oem
.\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
...
FOO=BAR
...
```
-> success

**Comment by yongtang** [comment]
@matt-richardson 
For UNICODE and UNICODE BigEndian: In Windows UNICODE is UTF16 and UNICODE BigEndian is UTF16BE so they are expected behavior.

For UTF7: UTF7 is technically a subset of ASCII and will encode `FOO=BAR` into `FOO+AD0-BAR`. Since there is no `=`, `FOO+AD0-BAR` will be considered as an empty env variable. This is consistent with Linux:
```
ubuntu@ubuntu:~/docker$ docker run -e "FOO+AD0-BAR" -e "NOFOO" -e "BAR=FOO" busybox env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=18437d11ffb0
BAR=FOO
HOME=/root
```

If  `FOO+AD0-BAR` is appended with `=` then env variable will show up:
```
ubuntu@ubuntu:~/docker$ docker run -e "FOO+AD0-BAR=FOOBAR" -e "NOFOO" -e "BAR=FOO" busybox env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=72ce28e8f9a5
FOO+AD0-BAR=FOOBAR
BAR=FOO
HOME=/root
```

**Comment by jhowardmsft** [comment]
Design LGTM

File: runconfig/opts/parse_test.go [hunk 1]
```diff
@@ -666,6 +666,39 @@ func TestParseEnvfileVariables(t *testing.T) {
...
+	env := []string{"FOO=BAR", "HELLO=" + string([]byte{0xe6, 0x82, 0xa8, 0xe5, 0xa5, 0xbd}), "BAR=FOO"}
+	if len(config.Env) != len(env) {
+		t.Fatalf("Expected a config with %d env variables, got %v: %v", len(env), len(config.Env), config.Env)
+	}
+	for i, v := range env {
+		if config.Env[i] != v {
+			t.Fatalf("Expected a config with [%s], got %v", v, []byte(config.Env[i]))
+		}
+	}
+	//if config.Env[1] != "HELLO=" {
```
Comment (jhowardmsft) [inline runconfig/opts/parse_test.go]: Should these lines be removed?

Comment (yongtang) [inline runconfig/opts/parse_test.go]: @jhowardmsft Sorry, didn't check carefully when pushed 😊  The PR has been fixed now. Thanks for the review.

**Comment by jhowardmsft** [comment]
CI failure looks like an unrelated blip. Restarted. 
LGTM if green this time

**Comment by yongtang** [comment]
@jhowardmsft I rebased the PR again and all tests passed. Please take a look and let me know if there are any issues.

**Comment by jhowardmsft** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Comment by thaJeztah** [comment]
Thanks everyone for helping!

**Comment by mazhar11-cou** [comment]
> ### Unicode (UTF16?)
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding unicode
> C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
> See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
> ```
> 
> -> Success, i think?
> 
> ### BigEndianUnicode
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding bigendianunicode
> C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
> See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
> ```
> 
> -> Success, i think?
> 
> ### UTF8
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf8
> .\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
> ...
> FOO=BAR
> ...
> ```
> 
> -> success
> 
> ### UTF7
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf7
> .\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
> ...
> ```
> 
> -> failure - no FOO variable
> 
> ### UTF32
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding utf32
> C:\Users\matt\Downloads\docker-1.13.0-dev.exe: env file env.list contains invalid utf8 bytes at line 1: [255 254 0 0 70 0 0 0 79 0 0 0 79 0 0 0 61 0 0 0 66 0 0 0 65 0 0 0 82 0 0 0 13 0 0 0].
> See 'C:\Users\matt\Downloads\docker-1.13.0-dev.exe run --help'.
> ```
> 
> -> Success, i think?
> 
> ### ASCII
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding ascii
> .\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
> ...
> FOO=BAR
> ...
> ```
> 
> -> success
> 
> ### OEM
> ```
> "FOO=BAR" | Out-File -FilePath .\env.list -Encoding oem
> .\docker-1.13.0-dev.exe run --env-file env.list --rm microsoft/windowsservercore cmd /c set
> ...
> FOO=BAR
> ...
> ```
> 
> -> success

This working correctly in my windows based PC

**Pull request merged**